File: stepik-union/src/main/java/org/stepik/core/metrics/Metrics.java
package org.stepik.core.metrics;

import com.intellij.openapi.application.ApplicationInfo;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;
import org.stepik.api.client.StepikApiClient;
import org.stepik.api.objects.metrics.Metric;
import org.stepik.api.queries.metrics.StepikMetricsPostQuery;
import org.stepik.core.StepikProjectManager;
import org.stepik.core.courseFormat.StepNode;
import org.stepik.core.courseFormat.StepType;
import org.stepik.core.courseFormat.StudyNode;
import org.stepik.core.utils.PluginUtils;
import org.stepik.core.utils.Utils;

import java.util.UUID;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

import static org.stepik.core.metrics.MetricsStatus.SUCCESSFUL;
import static org.stepik.core.stepik.StepikAuthManager.authAndGetStepikApiClient;
import static org.stepik.core.stepik.StepikAuthManager.isAuthenticated;

/**
 * @author meanmail
 */
public class Metrics {
    private static final Logger logger = Logger.getInstance(Metrics.class);
    private static final String session = UUID.randomUUID().toString();
    private static final ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);

    private static void postMetrics(
            @NotNull Project project,
            @NotNull Metric metric,
            @NotNull MetricsStatus status) {
        executor.schedule(() -> {
            StepikApiClient stepikApiClient;
            stepikApiClient = authAndGetStepikApiClient();
            if (!isAuthenticated()) {
                return;
            }

            final StepikMetricsPostQuery query = stepikApiClient.metrics()
                    .post()
                    .timestamp(System.currentTimeMillis() / 1000L)
                    .tags(metric.getTags())
                    .data(metric.getData())
                    .name("ide_plugin")
                    .tags("name", "S_Union")
                    .tags("ide_name", ApplicationInfo.getInstance().getVersionName())
                    .data("ide_version", ApplicationInfo.getInstance().getBuild().toString())
                    .data("plugin_version", PluginUtils.getVersion())
                    .data("session", session)
                    .tags("status", status);

            StepikProjectManager projectManager = StepikProjectManager.getInstance(project);

            if (projectManager != null) {
                query.data("project_id", projectManager.getUuid())
                        .tags("project_programming_language", projectManager.getDefaultLang().getLangName())
                        .data("project_manager_version", projectManager.getVersion());

                StudyNode projectRoot = projectManager.getProjectRoot();

                if (projectRoot != null) {
                    Class<? extends StudyNode> projectRootClass = projectRoot.getClass();
                    query.tags("project_root_class", projectRootClass.getSimpleName())
                            .data("project_root_id", projectRoot.getId());

                    query.data("course_id", projectRoot.getCourseId(stepikApiClient));
                }
            }

            query.executeAsync()
                    .exceptionally((e) -> {
                        String message = String.format("Failed post metric: %s", query.toString());
                        logger.warn(message, e);
                        return null;
                    });
        }, 500, TimeUnit.MILLISECONDS);
    }

    private static void postSimpleMetric(
            @NotNull Project project,
            @NotNull String action,
            @NotNull MetricsStatus status) {
        Metric metric = new Metric();
        metric.addTags("action", action);

        postMetrics(project, metric, status);
    }

    public static void authenticate(@NotNull MetricsStatus status) {
        Project project = Utils.getCurrentProject();
        postSimpleMetric(project, "authenticate", status);
    }

    public static void createProject(@NotNull Project project, @NotNull MetricsStatus status) {
        postSimpleMetric(project, "create_project", status);
    }

    public static void openProject(@NotNull Project project, @NotNull MetricsStatus status) {
        postSimpleMetric(project, "open_project", status);
    }

    private static void stepAction(
            @NotNull String actionName,
            @NotNull Project project,
            @NotNull StepNode stepNode,
            @NotNull MetricsStatus status) {
        Metric metric = new Metric();
        metric.addTags("action", actionName);
        metric.addData("step_id", stepNode.getId());
        metric.addTags("step_programming_language", stepNode.getCurrentLang().getLangName());
        StepType stepType = stepNode.getType();
        metric.addTags("step_type", stepType.getName());

        postMetrics(project, metric, status);
    }

    public static void sendAction(
            @NotNull Project project,
            @NotNull StepNode stepNode,
            @NotNull MetricsStatus status) {
        stepAction("send", project, stepNode, status);
    }

    public static void downloadAction(
            @NotNull Project project,
            @NotNull StepNode stepNode,
            @NotNull MetricsStatus status) {
        stepAction("download", project, stepNode, status);
    }

    public static void getStepStatusAction(
            @NotNull Project project,
            @NotNull StepNode stepNode,
            @NotNull MetricsStatus status) {
        stepAction("get_step_status", project, stepNode, status);
    }

    public static void insertAmbientCodeAction(
            @NotNull Project project,
            @NotNull StepNode stepNode,
            @NotNull MetricsStatus status) {
        stepAction("insert_ambient_code", project, stepNode, status);
    }

    public static void navigateAction(
            @NotNull Project project,
            @NotNull StudyNode studyNode,
            @NotNull MetricsStatus status) {
        if (studyNode instanceof StepNode) {
            StepNode stepNode = (StepNode) studyNode;
            stepAction("navigate", project, stepNode, status);
        }
    }

    public static void openInBrowserAction(
            @NotNull Project project,
            @NotNull StudyNode studyNode) {
        if (studyNode instanceof StepNode) {
            StepNode stepNode = (StepNode) studyNode;
            stepAction("open_in_browser", project, stepNode, SUCCESSFUL);
        }
    }

    public static void resetStepAction(
            @NotNull Project project,
            @NotNull StepNode stepNode,
            @NotNull MetricsStatus status) {
        stepAction("reset_step", project, stepNode, status);
    }

    public static void removeAmbientCodeAction(
            @NotNull Project project,
            @NotNull StepNode stepNode,
            @NotNull MetricsStatus status) {
        stepAction("remove_ambient_code", project, stepNode, status);
    }

    public static void switchLanguage(
            @NotNull Project project,
            @NotNull StepNode stepNode,
            @NotNull MetricsStatus status) {
        stepAction("switch_language", project, stepNode, status);
    }
}

File: stepik-union/src/main/kotlin/org/stepik/core/metrics/Metrics.kt
package org.stepik.core.metrics

import com.intellij.openapi.application.ApplicationInfo
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import org.stepik.api.client.StepikApiClient
import org.stepik.api.objects.metrics.Metric
import org.stepik.core.StepikProjectManager
import org.stepik.core.SupportedLanguages
import org.stepik.core.courseFormat.StepNode
import org.stepik.core.courseFormat.StudyNode
import org.stepik.core.metrics.MetricsStatus.SUCCESSFUL
import org.stepik.core.stepik.StepikAuthManager.authAndGetStepikApiClient
import org.stepik.core.stepik.StepikAuthManager.isAuthenticated
import org.stepik.core.utils.PluginUtils
import org.stepik.core.utils.Utils
import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit


object Metrics {
    private val logger = Logger.getInstance(Metrics::class.java)
    private val session = UUID.randomUUID().toString()
    private val executor = Executors.newScheduledThreadPool(1)

    private fun postMetrics(
            project: Project,
            metric: Metric,
            status: MetricsStatus) {
        executor.schedule({
            val stepikApiClient: StepikApiClient = authAndGetStepikApiClient()
            if (!isAuthenticated) {
                return@schedule
            }

            val query = stepikApiClient.metrics()
                    .post()
                    .timestamp(System.currentTimeMillis() / 1000L)
                    .tags(metric.tags)
                    .data(metric.data)
                    .name("ide_plugin")
                    .tags("name", "S_Union")
                    .tags("ide_name", ApplicationInfo.getInstance().versionName)
                    .data("ide_version", ApplicationInfo.getInstance().build.toString())
                    .data("plugin_version", PluginUtils.version)
                    .data("session", session)
                    .tags("status", status)

            val projectManager = StepikProjectManager.getInstance(project)

            if (projectManager != null) {
                query.data("project_id", projectManager.getUuid())
                        .tags("project_programming_language",
                                (projectManager.defaultLang ?: SupportedLanguages.INVALID).langName)
                        .data("project_manager_version", projectManager.version)

                val projectRoot = projectManager.projectRoot

                if (projectRoot != null) {
                    val projectRootClass = projectRoot.javaClass
                    query.tags("project_root_class", projectRootClass.simpleName)
                            .data("project_root_id", projectRoot.id)

                    query.data("course_id", projectRoot.getCourseId(stepikApiClient))
                }
            }

            query.executeAsync()
                    .exceptionally { e ->
                        val message = String.format("Failed post metric: %s", query.toString())
                        logger.warn(message, e)
                        null
                    }
        }, 500, TimeUnit.MILLISECONDS)
    }

    private fun postSimpleMetric(
            project: Project,
            action: String,
            status: MetricsStatus) {
        val metric = Metric()
        metric.addTags("action", action)

        postMetrics(project, metric, status)
    }

    fun authenticate(status: MetricsStatus) {
        val project = Utils.currentProject
        postSimpleMetric(project, "authenticate", status)
    }

    fun createProject(project: Project, status: MetricsStatus) {
        postSimpleMetric(project, "create_project", status)
    }

    fun openProject(project: Project, status: MetricsStatus) {
        postSimpleMetric(project, "open_project", status)
    }

    private fun stepAction(
            actionName: String,
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus) {
        val metric = Metric()
        metric.addTags("action", actionName)
        metric.addData("step_id", stepNode.id)
        metric.addTags("step_programming_language", stepNode.currentLang.langName)
        val stepType = stepNode.type
        metric.addTags("step_type", stepType.name)

        postMetrics(project, metric, status)
    }

    fun sendAction(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus) {
        stepAction("send", project, stepNode, status)
    }

    fun downloadAction(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus) {
        stepAction("download", project, stepNode, status)
    }

    fun getStepStatusAction(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus) {
        stepAction("get_step_status", project, stepNode, status)
    }

    fun insertAmbientCodeAction(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus) {
        stepAction("insert_ambient_code", project, stepNode, status)
    }

    fun navigateAction(
            project: Project,
            studyNode: StudyNode<*, *>,
            status: MetricsStatus) {
        if (studyNode is StepNode) {
            stepAction("navigate", project, studyNode, status)
        }
    }

    fun openInBrowserAction(
            project: Project,
            studyNode: StudyNode<*, *>) {
        if (studyNode is StepNode) {
            stepAction("open_in_browser", project, studyNode, SUCCESSFUL)
        }
    }

    fun resetStepAction(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus) {
        stepAction("reset_step", project, stepNode, status)
    }

    fun removeAmbientCodeAction(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus) {
        stepAction("remove_ambient_code", project, stepNode, status)
    }

    fun switchLanguage(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus) {
        stepAction("switch_language", project, stepNode, status)
    }

    fun testCodeAction(
            project: Project,
            stepNode: StepNode,
            status: MetricsStatus) {
        stepAction("test_code", project, stepNode, status)
    }
}
